### PR TITLE
Refactor jsg wrappable for oilpan

### PR DESCRIPTION
**Pull request by jasnell**
Updates the jsg Wrappable implementation around oilpan.

### CODE REVIEW

**Comment by jasnell** [comment]
TODOs on this PR:

* [ ] jsg/function.test.c++ failure
* [ ] jsg/jsg-test.c++ failure

**Review by harrishancock — COMMENTED** [review]
I'm partway through reviewing. Not done yet, but here's a tranche of (mostly non-substantive) comments.

File: src/workerd/jsg/wrappable.h [hunk 1]
```diff
@@ -8,255 +8,423 @@
...
   //
-  // Wrappable also maintains a *second* reference count on the wrapper itself. While the second
-  // refcount is non-zero, the wrapper (the JavaScript object) will not be allowed to be
-  // garbage-collected, even if there are no references to it from other JS objects. This is
-  // important if the C++ object may be re-exported to JavaScript in the future and needs to have
-  // the same identity at that point (including maintaining any monkey-patches that the script
-  // may have applied to it previously).
+  // Like a wrapper, a Ref is a handle to a Wrappable. A Ref can be either Strong or Traced.
+  // When a Ref is first created, it is always Strong -- that is, it holds a strong ref to
+  // the Wrappable that keeps it alive. If the Wrappable has a wrapper, a Strong ref *also*
```
[the remark this answers is not in the document]
Comment (jasnell) [reply src/workerd/jsg/wrappable.h:62]: the Wrappable, I'll clarify

File: src/workerd/jsg/resource.h [hunk 2]
```diff
@@ -950,7 +948,7 @@ class ResourceWrapper {
                                v8::NewStringType::kInternalized)));
 
     // Store a pointer to this object in slot 1, to be extracted in callbacks.
-    context->SetAlignedPointerInEmbedderData(1, ptr.get());
+    //context->SetAlignedPointerInEmbedderData(1, ptr.get());
```
[the remark this answers is not in the document]
Comment (jasnell) [reply src/workerd/jsg/resource.h:951]: ah right, yeah forgot about that one.

File: src/workerd/jsg/wrappable.h [hunk 3]
```diff
@@ -8,255 +8,423 @@
...
 class Wrappable: public kj::Refcounted {
-  // Base class for C++ objects which can be "wrapped" for JavaScript consumption. A JavaScript
-  // "wrapper" object is created, and then the JS wrapper and C++ Wrappable are "attached" to each
-  // other via attachWrapper().
+  // Base class for C++ objects which can be "wrapped" for JavaScript consumption.
+  // Wrappable is refcounted via kj::Refcounted. A single instance may have multiple
+  // handles keeping it alive. When the last handle is dropped, the Wrappable instance
+  // will be destroyed.
+  //
+  // The "wrapper" is a garbage collected handle for the Wrappable that has an associated
```
[the remark this answers is not in the document]
Comment (jasnell) [reply src/workerd/jsg/wrappable.h:37]: Technically it's the combination of `WrapperHandle`, `WrapperHandle::Shim`, and the JavaScript object. I'll clarify in the comment

File: src/workerd/jsg/wrappable.h [hunk 4]
```diff
@@ -8,255 +8,423 @@
...
-  //
-  // This handle always holds a v8::Object.
+  class WrapperHandle final{
+    // The WrapperHandle is used internally by Wrappable to manage the V8
+    // wrapper object (if one exists).
+  public:
+    class Shim;
+    // The Shim is an opaque object associated with the JS wrapper that holds
+    // a kj::Own reference to the Wrappable. It will stay alive as long as the
+    // wrapper stays alive.
```
[the remark this answers is not in the document]
Comment (jasnell) [reply src/workerd/jsg/wrappable.h:125]: I'll clarify

File: src/workerd/jsg/wrappable.h [hunk 5]
```diff
@@ -8,255 +8,423 @@
...
+  WrapperHandle wrapper;
+  // Handle to the JS wrapper object. The wrapper is created lazily when the object is first
+  // exported to JavaScript; until then, the wrapper is empty.
+
+  kj::Maybe<RefBase::StrongRefHandle&> maybeStrongRefHandle;
+  // There is only ever at most one refcounted StrongRefHandle created for a single
+  // Wrappable, with each "strong" RefBase holding a kj::Own reference to it. If
+  // the StrongRefHandle does not exist, one will be created on demand.
+
+  kj::Maybe<RefBase::TracedRefHandle&> maybeTracedRefHandle;
```
[the remark this answers is not in the document]
Comment (jasnell) [reply src/workerd/jsg/wrappable.h:300]: No, they can both be set. For instance,

```
Ref<Foo> foo = alloc<Foo>();
foo->child = foo.addRef();
```

When `foo` is traced, the `foo->child` Ref will be holding onto the `TracedRefHandle` while the original `foo` Ref will be holding onto the `StrongRefHandle`.

File: src/workerd/jsg/wrappable.h [hunk 6]
```diff
@@ -8,255 +8,423 @@
...
+    static kj::Maybe<Wrappable&> tryUnwrap(const v8::Local<v8::Context>& context,
+                                           const v8::Local<v8::Object>& object);
+    // If the given object is a proxy for a Wrappable, returns a reference to the
+    // instance. Otherwise returns nullptr.
+
+    template <typename T>
+    static kj::Maybe<T&> tryUnwrap(const v8::Local<v8::Context>& context,
+                                   const v8::Local<v8::Object>& object) {
+      KJ_IF_MAYBE(wrappable, tryUnwrap(context, object)) {
+        return reinterpret_cast<T&>(*wrappable);
```
[the remark this answers is not in the document]
Comment (jasnell) [reply src/workerd/jsg/wrappable.h:150]: I tried that originally but could get it to work. Kept getting compile errors.

e.g. 

```
bazel-out/k8-fastbuild/bin/external/capnp-cpp/src/kj/_virtual_includes/kj/kj/common.h:1910:10: error: cannot cast 'workerd::api::ExtendableEvent' to its private base class 'workerd::jsg::Wrappable'
  return kj::fwd<From>(from);
```

**Comment by jasnell** [comment]
Superseded by #199

**Pull request closed without merging**